ethylene-core/src/main/java/com/github/steanky/ethylene/core/AbstractConfigCodec.java

```diff
@@ -0,0 +1,192 @@
+package com.github.steanky.ethylene.core;
+
+import com.github.steanky.ethylene.core.collection.*;
+import org.jetbrains.annotations.Contract;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.*;
+
+/**
+ * <p>This class contains functionality common to many {@link ConfigCodec} implementations. Many of its methods are
+ * designed to deeply iterate hierarchical data structures.</p>
+ *
+ * <p>In addition to the two abstract methods, {@link AbstractConfigCodec#readObject(InputStream)} and
+ * {@link AbstractConfigCodec#writeObject(Object, OutputStream)}, almost every method is designed to be readily
+ * subclassed by implementations. In particular, specialized codecs may enable processing of custom objects that are not
+ * natively supported by this class.</p>
+ */
+public abstract class AbstractConfigCodec implements ConfigCodec {
+    /**
+     * Empty constructor, for use by subclasses.
+     */
+    protected AbstractConfigCodec() {}
+
+    @Override
+    public void encode(@NotNull ConfigElement element, @NotNull OutputStream output) throws IOException {
+        Objects.requireNonNull(element);
+        Objects.requireNonNull(output);
+
+        try (output) {
+            writeObject(
+                    GraphTransformer.process(element, this::makeEncodeNode, this::isContainer, this::serializeElement),
+                    output);
+        }
+    }
+
+    @Override
+    public @NotNull ConfigElement decode(@NotNull InputStream input) throws IOException {
+        Objects.requireNonNull(input);
+
+        try (input) {
+            return GraphTransformer.process(readObject(input), this::makeDecodeNode, this::isContainer,
+                    this::deserializeObject);
+        }
```

Comment:
I think it has always been like this but generally are you sure you want to be the one closing streams?